### PR TITLE
Allow to commit offset in case of deserialization errors

### PR DESCRIPTION
**Pull request by giamo**
# Problem

When a consumed message cannnot be deserialized (e.g. when using the Avro deserializer and receiving a JSON) the `DefaultKafkaListenerExceptionHandler` has an option to skip the message with a seek() and keep consuming the next message, but it doesn’t commit the message offset. This means that until a “good” message is processed and committed, the consumer group will show a lag and it will try reprocessing the bad message in case of restart or group rebalancing. This issue especially impacts low-traffic topics.

# Proposed solution

Add a flag `commitRecordOnDeserializationFailure` to indicate if the default error handler should also commit the offset in addition to doing the seek(). The flag is set to false by default, so the handler will continue with the current behavior unless the flag is set explicitly by the user.

### CODE REVIEW

**Review by sdelamo — CHANGES_REQUESTED** [review]
We would need a test if we wanted to go with this approach.

**Review by jeremyg484 — CHANGES_REQUESTED** [review]
Logically I think this option makes sense overall. I'd potentially like to see more fine-grained error handling as mentioned, and as @sdelamo said, it needs tests.

**Review by guillermocalvo — CHANGES_REQUESTED** [review]
LGTM except for the lack of tests.

As an altenative approach, we could have merged those two boolean flags into a single configurable property `onDeserializationFailure` or `deserializationFailureStrategy` that could be populated with an enum type like this:

```java
public enum DeserializationFailureStrategy {
    /** Does nothing. */
    DO_NOTHING,

    /** Seeks past the offending record. */
    SKIP,

    /** Skips and commits offending record. */
    COMMIT
}
```

But I admit the current approach is more backward compatible.

File: kafka/src/main/java/io/micronaut/configuration/kafka/exceptions/DefaultKafkaListenerExceptionHandler.java [hunk 1]
```diff
@@ -95,9 +106,13 @@ protected void seekPastDeserializationError(
                 TopicPartition tp = new TopicPartition(topic, partition);
                 LOG.debug("Seeking past unserializable consumer record for partition {}-{} and offset {}", topic, partition, offset);
                 kafkaConsumer.seek(tp, offset + 1);
+                if (this.commitRecordOnDeserializationFailure) {
```
[the remark this answers is not in the document]
Comment (giamo) [reply kafka/src/main/java/io/micronaut/configuration/kafka/exceptions/DefaultKafkaListenerExceptionHandler.java:109]: I guess there should be three separate try/catch then… one for parsing the error message, one for the seek and one for the commit

**Comment by giamo** [comment]
> LGTM except for the lack of tests.
> 
> As an altenative approach, we could have merged those two boolean flags into a single configurable property `onDeserializationFailure` or `deserializationFailureStrategy` that could be populated with an enum type like this:
> 
> ```java
> public enum DeserializationFailureStrategy {
>     /** Does nothing. */
>     DO_NOTHING,
> 
>     /** Seeks past the offending record. */
>     SKIP,
> 
>     /** Skips and commits offending record. */
>     COMMIT
> }
> ```
> 
> But I admit the current approach is more backward compatible.

Then enum would make sense but I suppose it’s better to be backward compatible and not change the interface and the behavior for people already using the handler. 

I will try to add the test one of these days!

**Comment by giamo** [comment]
@sdelamo @jeremyg484 @guillermocalvo added a test and a separate try/catch for the commit operation

**Comment by sdelamo** [comment]
> Then enum would make sense but I suppose it’s better to be backward compatible and not change the interface and the behavior for people already using the handler.

Yes, we cannot introduce a breaking change.

File: kafka/src/main/java/io/micronaut/configuration/kafka/exceptions/DefaultKafkaListenerExceptionHandler.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (sdelamo) [reply kafka/src/main/java/io/micronaut/configuration/kafka/exceptions/DefaultKafkaListenerExceptionHandler.java:109]: can you do the three separate try/catch @giamo ?

**Review by guillermocalvo — APPROVED** [review]

File: kafka/src/main/java/io/micronaut/configuration/kafka/exceptions/DefaultKafkaListenerExceptionHandler.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (giamo) [reply kafka/src/main/java/io/micronaut/configuration/kafka/exceptions/DefaultKafkaListenerExceptionHandler.java:116]: done @sdelamo

**Review by jeremyg484 — APPROVED** [review]

File: kafka/src/test/groovy/io/micronaut/configuration/kafka/exceptions/DefaultKafkaListenerExceptionHandlerSpec.groovy [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+        DefaultBehaviorOnDeserializationErrorConsumer() {
+            super(TOPIC_SEEK)
+        }
+
+        @Topic(TOPIC_SEEK)
+        void receive(UUID uuid) {
+        }
+    }
+
+    @KafkaListener(
```
Comment (sdelamo) [inline kafka/src/test/groovy/io/micronaut/configuration/kafka/exceptions/DefaultKafkaListenerExceptionHandlerSpec.groovy:110]: this class needs to be annotated with ` @Requires(property = 'spec.name',` as well.

**Review by sdelamo — CHANGES_REQUESTED** [review]
How will people discover this new settings? I think we have to add docs.

File: kafka/src/test/groovy/io/micronaut/configuration/kafka/exceptions/DefaultKafkaListenerExceptionHandlerSpec.groovy [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+        given:
+        StringProducer stringProducer = context.getBean(StringProducer)
+        DoNothingOnDeserializationErrorConsumer consumer = context.getBean(DoNothingOnDeserializationErrorConsumer)
+
+        when: "A producer sends a message with wrong serialization"
+        stringProducer.sendToDoNothingTopic("not-a-uuid")
+
+        then: "The message is skipped, but not committed"
+        conditions.eventually {
+            consumer.currentPosition > 0
```
Comment (giamo) [inline kafka/src/test/groovy/io/micronaut/configuration/kafka/exceptions/DefaultKafkaListenerExceptionHandlerSpec.groovy:62]: @guillermocalvo it's strange that this test passes, the `DoNothingOnDeserializationErrorConsumer` explicitly sets the `skipRecordOnDeserializationFailure` to false so I'd expect the offset not to be advanced... or am I missing something?

Comment (guillermocalvo) [reply kafka/src/test/groovy/io/micronaut/configuration/kafka/exceptions/DefaultKafkaListenerExceptionHandlerSpec.groovy:62]: @giamo Previous behavior was fixed by: https://github.com/micronaut-projects/micronaut-kafka/pull/771

File: src/main/docs/guide/kafkaListener/kafkaErrors.adoc [hunk 4]
```diff
@@ -46,6 +46,14 @@ NOTE: Specify exception to retry apply only for RETRY_ON_ERROR error strategy.
 
 When an exception occurs in a ann:configuration.kafka.annotation.KafkaListener[] method by default the exception is simply logged. This is handled by api:configuration.kafka.exceptions.DefaultKafkaListenerExceptionHandler[].
 
+[TIP]
+====
+DefaultKafkaListenerExceptionHandler behavior can be configured via methods:
+
+- `setSkipRecordOnDeserializationFailure`: Sets whether the seek past records that are not deserializable (enabled by default).
```
Comment (sdelamo) [inline src/main/docs/guide/kafkaListener/kafkaErrors.adoc:53]: I think we should create a configuration object and let users configure this in `application.properties`

**Review by sdelamo — CHANGES_REQUESTED** [review]

File: src/main/docs/guide/kafkaListener/kafkaErrors.adoc (hunk 4 above)
Comment (guillermocalvo) [reply src/main/docs/guide/kafkaListener/kafkaErrors.adoc]: @sdelamo Done 366528a1b68b2b7647c52e6542884b5391d7dadc

File: kafka/src/main/java/io/micronaut/configuration/kafka/exceptions/DefaultKafkaListenerExceptionHandler.java [hunk 5]
```diff
@@ -42,7 +46,19 @@ public class DefaultKafkaListenerExceptionHandler implements KafkaListenerExcept
...
-    private boolean skipRecordOnDeserializationFailure = true;
+    private boolean skipRecordOnDeserializationFailure;
+    private boolean commitRecordOnDeserializationFailure;
+
+    /**
+     * Creates a new instance.
+     *
+     * @param config The default Kafka listener exception handler configuration
+     */
+    @Inject
```
Comment (sdelamo) [inline kafka/src/main/java/io/micronaut/configuration/kafka/exceptions/DefaultKafkaListenerExceptionHandler.java:57]: add an empty constructor and annotate with `@Deprecated`

**Review by sdelamo — CHANGES_REQUESTED** [review]

**Review by sdelamo — APPROVED** [review]

**Pull request merged**